### PR TITLE
Deprecating requiresAuthentication Attribute on SFRequest

### PR DESCRIPTION
**Pull request by trooper2013**
Moving forward unauthenticated requests should be sent using [SFRestAPI sharedGlobalInstance]  
and requests requiring authentication context(accessToken) should  use [SFRestAPI sharedInstance:..].

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 1]
```diff
@@ -209,18 +210,22 @@ + (NSString *)userAgentString:(NSString*)qualifier {
...
 }
 
 - (void)send:(SFRestRequest *)request delegate:(id<SFRestDelegate>)delegate shouldRetry:(BOOL)shouldRetry {
     if (nil != delegate) {
         request.delegate = delegate;
     }
-
+    
+    SFSDK_USE_DEPRECATED_BEGIN
+        NSAssert(!request.requiresAuthentication, @"Use SFRestAPI, sharedGlobalInstance for unauthenticated requests or sharedInstance for authenticated requests");
```
[the remark this answers is not in the document]
Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: I see you want to make sure it's not used anymore.
What about making sure (un)authenticated requests go through (un)authenticated rest client?

**Comment by wmathurin** [comment]
I'm having second thoughts about this. Before the requests knew what it needed. Take RestAPI explorer for instance, on one side we build requests, on the other we send them. How is the sending code going to know through what SFRestAPI the request should be sent? Or you could have a custom sync target going to Heroku, how is the send method of the SyncManager know what to do with the request. I think we need to keep that flag and simply check that you don't try to send a request requiring authentication through the unauthenticated SFRestAPI.

**Comment by wmathurin** [comment]
The global SFRestAPI should just be used before you have logged in. Once you have logged in you can use the other one and control with each request whether you want the oauth token header or not.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 2]
```diff
@@ -410,7 +414,7 @@ - (void)resendActiveRequestsRequiringAuthentication {
     @synchronized (self) {
         NSSet *pendingRequests = [self.activeRequests asSet];
         for (SFRestRequest *request in pendingRequests) {
-            if (request.requiresAuthentication) {
+            if (self.requiresAuthentication) {
```
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: @wmathurin Thats correct. I don't think we need this check. Will remove it.

File: libs/SalesforceSDKCore/SalesforceSDKCoreTests/SalesforceRestAPITests.m [hunk 3]
```diff
@@ -149,6 +149,13 @@ - (void)testGetVersions {
     XCTAssertEqualObjects(listener.returnStatus, kTestRequestStatusDidLoad, @"request failed");
 }
 
+- (void)testGetVersionsUnAuthenticated {
+    SFRestRequest* request = [[SFRestAPI sharedInstance] requestForVersions];
+    request.requiresAuthentication = NO;
```
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCoreTests/SalesforceRestAPITests.m]: @wmathurin  SFRestAPI:requestForVersions did return a SFRequest with requiredAuthentication as false, I will need to add that back in to the factory method.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 4]
```diff
@@ -48,7 +48,7 @@ @interface SFRestAPI ()
 @property (readwrite, assign) BOOL pendingRequestsBeingProcessed;
 @property (nonatomic, strong) SFOAuthSessionRefresher *oauthSessionRefresher;
 @property (nonatomic, strong, readwrite) SFUserAccount *user;
-
+@property (nonatomic, assign) BOOL requiresAuthentication;
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: Should we call that something else like isAuthenticated?

**Pull request merged**